### PR TITLE
Gh test

### PR DESCRIPTION
**Pull request by nilesh05apr**
Implement a github action to check unit tests #33

Fresh PR for the above task.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        os: [ubuntu-latest, windows-latest]
+        python-version: ['3.8', '3.9', '3.10']
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Install dependencies
```
Comment (sudiptob2) [inline .github/workflows/test.yml:43]: we need to cache the dependencies for a faster run. Can you do some research and implement this caching mechanism, please?
Here is an example : [caching dep example](https://github.com/sudiptob2/cf-stats/blob/befabb28dfb41e34a5151fad1898842152c3cb33/.github/workflows/test.yml#L23)

**Review by sudiptob2 — COMMENTED** [review]
We need several improvements in this action, Let's work on it step by step. :blush:

**Review by sudiptob2 — COMMENTED** [review]
Thanks :pray: 
Now we have to run `check.sh` script instead of running `pytest` directly in the workflow. I have implemented the required checks in that script, please take a look.

Also, I could not check the cache hit, could you please double-check and make sure caching is working :blush:

**Comment by nilesh05apr** [comment]
@sudiptob2 i have manually re checked, all jobs are running fine and cache hit also successful. Now i need to work on running the bash script in the workflow.
Please check and review the progress.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        run: |
+          python3 -m pip install --upgrade pip setuptools wheel
+          python3 -m pip install -r requirements.txt
+          pip install -e .
+      # Run tests
+      - name: Test
+        run: |
+          python3 --version
+          python3 -m pip install pytest pytest-mock
+          python3 -m pytest
```
Comment (sudiptob2) [inline .github/workflows/test.yml]:
```suggestion
```

**Review by sudiptob2 — CHANGES_REQUESTED** [review]
we don't need to run pytest, its already inside the `check.sh`

**Comment by sudiptob2** [comment]
@nilesh05apr I think if you could use `pipenv` and install dependencies using `Pipfile` then you need not to install these `dev dependencies` manually.
Because these dev dependencies are not mentioned in the `requirements.txt`, it's a hassle to install them one by one in the `ghithub actions` manually.
if you use `pipenv` then you can just do
```
run: pipenv install --dev --verbose
```
see this [github action](https://github.com/chkware/cli/blob/main/.github/workflows/linux-run-tests.yml) for example.

**Comment by nilesh05apr** [comment]
@sudiptob2 can you please suggest any fix? now caching is also failing.

**Comment by sudiptob2** [comment]
One thing I can understand about caching is that the cache key is not right. As you are using a matrix of python, you have to append the python version to make those cache keys unique. Also, you have to remove windows OS, because things are different there.
Apart from that remove these lines
```bash
      # Run tests
      - name: Test
        run: |
          python3 --version
          python3 -m pip install pytest pytest-mock flake8 mypy requests
          python3 -m pytest 
```
these are not necessary because, the `check.sh` already doing these tests.

**Comment by nilesh05apr** [comment]
@sudiptob2  i updated the code but the problem for check.sh remains same.

**Review by sudiptob2 — COMMENTED** [review]
seems better now, But we still have errors right?
below checks might help.
**caching** : run the action multiple time and confirm that the cache is working
**check.sh**: I can see this script is failing, I suppose, I was using `python` as interpreter may be `python3` is needed. Also maybe we need to install the missing stubs

**Comment by nilesh05apr** [comment]
for the caching part i will re run the jobs as many times as required but for the check.sh file i am unable to find and helpful resource.

**Review by sudiptob2 — COMMENTED** [review]
In my opinion, these changes are goin to the wrong direction, your previous commit was OK, just it might need to install some stub packages, you can revert these changes and I will try to check it myself and you know If I can fix it.

**Comment by nilesh05apr** [comment]
the problem is caching is also not working properly i have checked several times, can you please comment?

**Comment by sudiptob2** [comment]
> the problem is caching is also not working properly i have checked several times, can you please comment?

Ok, just keep this PR at the previous commit, I will test and update you if I can find anything. You can always do experiments in a separate branch right.

**Comment by nilesh05apr** [comment]
sure that i can do, i will create a new branch and try experimenting. Thanks

**Comment by sudiptob2** [comment]
I have implemented/fixed the workflow, see [here](https://github.com/sudiptob2/leet-easy/blob/experiment/action/.github/workflows/test.yml)

you can update your PR with the [this code](https://github.com/sudiptob2/leet-easy/blob/experiment/action/.github/workflows/test.yml)

**Comment by nilesh05apr** [comment]
Thanks so much @sudiptob2 for the help. I have implemented the changes and have ensured everything is running smoothly.

**Comment by nilesh05apr** [comment]
@ can you please review and merge the pull request. Thank you

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        os: [ubuntu-latest, macOS-latest]
+        python-version: ['3.8', '3.9', '3.10']
+
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      - name: Python version
```
Comment (sudiptob2) [inline .github/workflows/test.yml:23]: add a space before this line

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
+name: Test
+on:
+  - pull_request
```
Comment (sudiptob2) [inline .github/workflows/test.yml]:
```suggestion
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  - push
```
Comment (sudiptob2) [inline .github/workflows/test.yml]:
```suggestion
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
```
Comment (sudiptob2) [inline .github/workflows/test.yml:5]:
```suggestion
  pull_request:
    types: [ready_for_review]
```

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      # create .venv folder
+      - name: create .venv folder
+        run: mkdir -p .venv
+
+      # caching dependencies
+      - name: Caching Dependencies
+        uses: actions/cache@v2
+        id: cache-dependencies
+        with:
+          path: ./.venv
```
Comment (sudiptob2) [inline .github/workflows/test.yml]:
```suggestion
          path: .venv
```

**Review by sudiptob2 — COMMENTED** [review]
There are a few small differences from what I implemented [here](https://github.com/sudiptob2/leet-easy/blob/experiment/action/.github/workflows/test.yml)
I might have miss a few thing while reviewing this, I would suggest you copy and paste the entire code from [here](https://github.com/sudiptob2/leet-easy/blob/experiment/action/.github/workflows/test.yml) to your file

**Comment by nilesh05apr** [comment]
@sudiptob2 i have copy-pasted the code XD. please review the changes

**Review by sudiptob2 — APPROVED** [review]
seems OK

**Comment by nilesh05apr** [comment]
Thank you @sudiptob2 i had great experience working on this issue.

**Pull request merged**